subsys/bluetooth/host/scan.c

```diff
@@ -620,6 +620,24 @@ void bt_hci_le_adv_ext_report(struct net_buf *buf)
 		is_report_complete = data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE;
 		more_to_come = data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_PARTIAL;
 
+		if (evt->length > buf->len) {
+			LOG_WRN("Adv report corrupted (wants %u out of %u)", evt->length, buf->len);
+
+			net_buf_reset(buf);
+
+			if (evt_type & BT_HCI_LE_ADV_EVT_TYPE_LEGACY) {
+				return;
+			}
+
+			/* Start discarding irrespective of the `more_to_come` flag. We
+			 * assume we may have lost a partial adv report in the truncated
+			 * data.
+			 */
+			reassembling_advertiser.state = FRAG_ADV_DISCARDING;
```

Comment:
I think the early return may cause us to produce garbled advertisement reports.

>  Start discarding irrespective of the `more_to_come` flag. We assume we may have lost a partial adv report in the truncated data.

The above comment is there because we are in a while-loop over reports. By returning early, we may be skipping over one or more reports. This can lead to a bad reassembly.

A bad reassembly is not a huge problem. The application has to tolerate malicious advertisements anyway. But when we did discussed this last time, we decided to prefer dropping data.

Reply:
can we come up with a test case for this somehow?

Reply:
but in this particular case, it's the controller that's malicious, no? so likely not a problem for "normal" use-cases. This early-return is more for protecting the host (and its CPU / resources I guess). My opinion is that we should just shut down bluetooth when that happens. It indicates something is wrong, either HCI driver synchronization or that the controller is very buggy or compromised.

Reply:
Yes. This is only relevant for HCI protocol violation. Either the controller did bad, or there was data corruption during HCI transport. It's not an exposed attack surface.